File: 刷题/cpp/46. 全排列.cpp
class Solution {
private:
    vector<vector<int>> ans;
public:
    void dfs(vector<int>& nums,vector<int>& tmp,int ind,int n,unordered_map<int,bool> used){
        if(ind==n){
            ans.push_back(tmp);
            return;
        }
        for(int i=0;i<n;i++){
            if(!used[nums[i]]){
                tmp.push_back(nums[i]);
                used[nums[i]]=true;
                dfs(nums,tmp,ind+1,n,used);
                tmp.pop_back();
                used[nums[i]]=false;
            }
            
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        int n=nums.size();
        unordered_map<int,bool> used;
        vector<int> tmp;
        dfs(nums,tmp,0,n,used);
        return ans;
    }
};

File: 刷题/golang/46. 全排列.go
func permute(nums []int) [][]int {
    n := len(nums)
    var ans [][]int
    var dfs func(int)
    dfs = func(ind int) {
        if ind == n {
            ans = append(ans, append([]int(nil), nums...))
            return
        }
        for i := ind; i < n; i++ {
            nums[i], nums[ind] = nums[ind], nums[i]
            dfs(ind+1)
            nums[ind], nums[i] = nums[i], nums[ind]
        } 
    }
    dfs(0)
    return ans
}